### DISCUSSION TITLE
Trouble Accessing Java Classes from Python Using GraalPy and GraalVM

### DISCUSSION THREAD

**Discussion by ShuzoShinagawa1102** [oracle/graalpython#448] [Q&A] [answered]
Hello everyone,

### What I want to do
I'm trying to access a Java class from a Python script using GraalPy (GraalPython) with GraalVM, but I'm encountering an error that I can't seem to resolve.

**Python Script (main.py)**
```
import java

# Get the Java classpath and print it
System = java.type("java.lang.System")
classpath = System.getProperty("java.class.path")
print(f"Java Classpath: {classpath}")

# Attempt to access the Java class
User = java.type('com.example.model.User')

# Create an instance of the Java class
user = User('Alice')
print(user.getName())
```

**Java Class(User.java)**
```
package com.example.model;

public class User {

    private String name;

    public User(String name) {
        this.name = name;
    }

    public String getName() {
        return this.name;
    }
}

```

I've compiled the Java class and packaged it into a JAR file named gpdemo-0.0.1-SNAPSHOT-plain.jar, and I set it as classpath for zshrc.

**Execution command**
```
graalpy --jvm \
  --vm-cp=$CLASSPATH \
  ./main.py

```

**Error Message:**
```
User = java.type('com.example.model.User')
KeyError: 'host symbol com.example.model.User is not defined or access has been denied'
````

### What I've Tried:
- Verified that the JAR file is in the correct location and that the class is properly packaged.
- Printed out the Java classpath within the Python script to ensure that the JAR file is included.
- Ensured that GraalVM is installed correctly and that I'm using the --jvm option to enable JVM interop.

### Environment Details:
GraalVM Version: Oracle GraalVM 23.0.1+11.1
GraalPy Version: GraalPy 3.10.8 (GraalVM CE JVM 23.1.2)
Operating System: macOS

### Questions:
- Is there something I'm missing in the way I'm specifying the classpath when running the script?
- Do I need to adjust any settings related to host access or permissions to allow Python to access the Java class?
- Has anyone else encountered a similar issue and found a solution?

**Comment by ShuzoShinagawa1102** [comment]
additionally, here is the contained classes which are showed by 
``` jar tf $CLASSPATH  ```
command.


```
MacBook-Pro-4 gpdemo % jar tf $CLASSPATH
META-INF/
META-INF/MANIFEST.MF
com/
com/example/
com/example/demo/
com/example/demo/GpdemoApplication.class
com/example/model/
com/example/model/User.class
application.properties
```

**Comment by msimacek** [comment]
Hi @ShuzoShinagawa1102, the option for supplying classpath when running through the graalpy launcher is `--vm.cp=$CLASSPATH` (not `-vm-cp`).

**Reply by ShuzoShinagawa1102** [reply]
@msimacek Thank you for pointing that out. It was just a typo on my part, and I am indeed using "--vm.cp" when runnng the command.

**Reply by msimacek** [reply, marked as answer]
I tried your example and it printed `Alice` with no errors. What I did:
- Save your java code as `com/example/model/User.java` and your python code as `main.py`
- `javac com/example/model/User.java`
- `graalpy --vm.cp=. main.py`

The differences I have from you is that I used the latest release, I'm on Linux and I didn't build a JAR. Could you please try the latest release ([24.1.1](https://github.com/oracle/graalpython/releases/tag/graal-24.1.1), the `graalpy-jvm` variant)?

**Reply by ShuzoShinagawa1102** [reply]
@msimacek 
Thank you for your kind response.

I installed GraalPy latest release version (24.1.1) that you suggested (https://github.com/oracle/graalpython/releases) and it worked.
The command which I executed is here.
```
graalpy --jvm \
  --vm.cp=gpdemo-0.0.1-SNAPSHOT-plain.jar \
  ./main.py
```
it returned ↓
<img width="410" alt="スクリーンショット 2024-10-30 22 54 57" src="https://github.com/user-attachments/assets/07f82a32-edb8-4943-a26d-4cfb2e85b700">
